agents/postgres/pgstatmonitor/pgstatmonitor.go

```diff
@@ -0,0 +1,302 @@
+	t := time.NewTimer(wait)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.changes <- agents.Change{Status: inventorypb.AgentStatus_STOPPING}
+			m.l.Infof("Context canceled.")
+			return
+
+		case <-t.C:
+			if !running {
+				m.changes <- agents.Change{Status: inventorypb.AgentStatus_STARTING}
+			}
+
+			lengthS := uint32(math.Round(wait.Seconds())) // round 59.9s/60.1s to 60s
+			buckets, err := m.getNewBuckets(ctx, start, lengthS)
+
+			start = time.Now()
+			wait = start.Truncate(queryStatMonitor).Add(queryStatMonitor).Sub(start)
+			m.l.Debugf("Scheduling next collection in %s at %s.", wait, start.Add(wait).Format("15:04:05"))
+			t.Reset(wait)
+
+			if err != nil {
+				m.l.Error(err)
+				running = false
+				m.changes <- agents.Change{Status: inventorypb.AgentStatus_WAITING}
+				continue
+			}
+
+			if !running {
+				running = true
+				m.changes <- agents.Change{Status: inventorypb.AgentStatus_RUNNING}
+			}
+
+			m.changes <- agents.Change{MetricsBucket: buckets}
+		}
+	}
+}
+
+func (m *PGStatMonitorQAN) getNewBuckets(ctx context.Context, periodStart time.Time, periodLengthSecs uint32) ([]*agentpb.MetricsBucket, error) {
+	current, prev, err := m.monitorCache.getStatMonitorExtended(ctx, m.q)
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := makeBuckets(current, prev, m.disableQueryExamples, m.pgsmNormalizedQuery, m.l)
+	startS := uint32(periodStart.Unix())
+	m.l.Debugf("Made %d buckets out of %d stat monitor in %s+%d interval.",
+		len(buckets), len(current), periodStart.Format("15:04:05"), periodLengthSecs)
+
+	// merge prev and current in cache
+	m.monitorCache.refresh(current)
+	m.l.Debugf("statMonitorCache: %s", m.monitorCache.stats())
+
+	// add agent_id and timestamps
+	for i, b := range buckets {
+		b.Common.AgentId = m.agentID
+		b.Common.PeriodStartUnixSecs = startS
+		b.Common.PeriodLengthSecs = periodLengthSecs
+
+		buckets[i] = b
+	}
+
+	return buckets, nil
+}
+
+// makeBuckets uses current state of pg_stat_monitor table and accumulated previous state
+// to make metrics buckets.
+//
+// makeBuckets is a pure function for easier testing.
+func makeBuckets(current, prev map[string]*pgStatMonitorExtended, disableQueryExamples, pgsmNormalizedQuery bool, l *logrus.Entry) []*agentpb.MetricsBucket {
+	res := make([]*agentpb.MetricsBucket, 0, len(current))
+
+	for queryID, currentPSS := range current {
+		prevPSS := prev[queryID]
+		if prevPSS == nil {
+			prevPSS = new(pgStatMonitorExtended)
+		}
+		count := float32(currentPSS.Calls - prevPSS.Calls)
+		switch {
+		case count == 0:
+			// Another way how this is possible is if pg_stat_monitor was truncated,
+			// and then the same number of queries were made.
+			// Currently, we can't differentiate between those situations.
+			l.Tracef("Skipped due to the same number of queries: %s.", currentPSS)
+			continue
+		case count < 0 || (time.Now().Sub(currentPSS.BucketStartTime).Seconds() > 60):
```

Comment:
Why we need that condition there, but not for pg_stat_statements?

Reply:
I am not sure why it is not in pg_stat_statements too, but Nurlan highlight that for this can there should be also condition for start bucket time. I checked interval and add this condition with more than 60 seconds to create new bucket.